### PR TITLE
Scaffold: Properly handle action's height being larger than available content height

### PR DESCRIPTION
**Pull request by shanio**
#504

### CODE REVIEW

**Review by Drekorian — APPROVED** [review]

**Review by hrach — CHANGES_REQUESTED** [review]
There are some visual glitches as can be seen here:
obrazek


Also, practically, the action slot should be probably limited to something like 50% of the available space to keep the content practically accessible. Covering the "content" by another "content" should be done differently  - instead of misusing the action slot.

I'll double-check the actual changes on Monday.

**Comment by shanio** [comment]
@hrach The changes applied are only fixing the underlying issues in height calculation when the action is too big. The space given to the content is deliberately not capped at the moment (similar to the original state), however the nature of the action slot is that it leaves 8.dp for content in form of gradient.

"The glitches" are actually parts of content, that is being shrunk to 8.dp, which is imo a little bit unexpected (tha action appears to the user as if it is above the content, not sharing/ taking away space from it. )

A fix size/ratio/weight could be applied as suggested, however it might not be so obvious, what would be a good value there. Also the problem will not be fixed by this completely. We will need to either disallow placing custom content into action slot or make sure that scrolling is applied in usage sites.

TL;DR: Making the action slot smaller needs to be coupled with fixes on the usage sites to ensure they fits, while this fix alone preserves the status quo of the component, it just does not allow to crash in those situations.

File: ui/src/androidMain/kotlin/kiwi/orbit/compose/ui/controls/Scaffold.kt [hunk 1]
```diff
@@ -91,31 +93,27 @@ private fun ScaffoldLayout(
...
+        val actionPlaceables = subcompose(SlotAction, action).map { it.measure(actionConstraints) }
         val actionHeight = actionPlaceables.maxOfOrNull { it.height } ?: 0
         val actionFadeHeight = actionPlaceables.firstOrNull()?.get(ActionFadeLine)?.let { value ->
             if (value == AlignmentLine.Unspecified) 0 else value
         } ?: 0
 
         val contentInsets = contentWindowInsets.asPaddingValues(this)
-        val contentTop = topBarPlaceables.maxOfOrNull { it.height } ?: 0
-        val contentBottom = if (actionHeight > 0) {
+        val occupiedBottomSize = if (actionHeight > 0) {
```
Comment (hrach) [inline ui/src/androidMain/kotlin/kiwi/orbit/compose/ui/controls/Scaffold.kt:107]: Please keep the naming - e.g. `contentBottom` here. This way it is inconsistent - size is actually height, and occupied is put here but not to other vars. Similarly, keep the contentTop please instead of topBarHeight.

File: ui/src/androidMain/kotlin/kiwi/orbit/compose/ui/controls/Scaffold.kt [hunk 2]
```diff
@@ -174,15 +174,17 @@ public fun ScaffoldAction(
         },
     ) { measurables, constraints ->
         val padding = 16.dp.roundToPx()
+        val top = fadeHeight.roundToPx()
         val action = measurables.first().measure(
-            constraints.offset(horizontal = padding * -2),
+            constraints
+                .offset(horizontal = padding * -2)
+                .copy(maxHeight = (constraints.maxHeight - padding - top).coerceAtLeast(0)),
```
Comment (hrach) [inline ui/src/androidMain/kotlin/kiwi/orbit/compose/ui/controls/Scaffold.kt:181]: using offset(vertical) is probably better as it accounts for max sizes and do the coercing for you:
```kotlin
                .offset(
                    horizontal = padding * -2,
                    vertical = -padding - top,
                ),
```

File: ui/src/androidMain/kotlin/kiwi/orbit/compose/ui/controls/Scaffold.kt (hunk 1 above)
Comment (shanio) [reply ui/src/androidMain/kotlin/kiwi/orbit/compose/ui/controls/Scaffold.kt:107]: Changed and renamed/preserved based on our discussion.

**Review by hrach — APPROVED** [review]
LGTM, screenshots have to be updated
